Add a -j/--json option for euca-describe-* commands

minnus opened the pull request:
Outputs json format instead of the standard text formatting
for these tools.

Addresses some of : TOOLS-626

Not sure if the arg parsing should go in **init** but it was the easiest way to get this going instead of adding it to each subcommand.

minnus committed:
Add a -j/--json option for euca-describe-* commands

Outputs json format instead of the standard text formatting
for these tools.

gholms commented:
This is an interesting idea!  If the option is going to go into `__init__` then it would probably make more sense to put format selection there, too, which would eliminate nearly all of the code duplication in this patch.  One way of doing that would be to rename `print_result` to something else in the affected commands and then add a new `print_result` method to EC2Request that decides whether to output JSON or call the renamed methods.  That's a bit more work, but it would get us a bit closer to having pluggable output formats while not having to hack up or override the `run` method.

I guess I should ask, though:  what do you want to accomplish with JSON-formatted output?  If we're going to be adding alternative output formats then we might want a more extensible way of invoking them than a dedicated option for each one.  The `--pretty=yaml` option in the euserv-describe-events tool that recently hit mainline eucalyptus would be a good example of that.

One other thing:  all euca2ools changes that affect the command line interface must come with documentation.  Would you please update the relevant man pages to reflect your changes?

gholms commented on euca2ools/commands/ec2/describeaccountattributes.py:
```diff
@@ -27,6 +27,7 @@
 
 from euca2ools.commands.ec2 import EC2Request
 
+import json
```
Standard library imports should come before all other imports, per PEP8.

minnus replied:
OK, Done

gholms commented on euca2ools/commands/ec2/describeaccountattributes.py:
```diff
@@ -35,6 +36,9 @@ class DescribeAccountAttributes(EC2Request):
     LIST_TAGS = ['accountAttributeSet', 'attributeValueSet']
 
     def print_result(self, result):
+        if self.args['json']:
```
Requests can invoke others by passing args directly to their initializers.  This means `json` may not always appear in self.args, so these checks should use `self.args.get('json')` instead.

This is generally less problematic in print_result, but it is useful to be consistent.

minnus replied:
OK, Done

minnus committed:
Move format selection to __init__

Make the code simpler by moving configuration to __init__ and
renaming the subclass print_result method.

minnus commented on euca2ools/commands/ec2/__init__.py:
```diff
@@ -72,6 +72,12 @@ class EC2Request(AWSQueryRequest, TabifyingMixin):
+            Arg('-j', '--json',
+                action='store_const', const='true', route_to=None,
+                help='''Output in json format''')
+            ]
```
OK, Done

minnus committed:
Use args.get for json parameter

Per gholms, json may not always appear in self.args since
requests can invoke others by passing args directly to their
initializers.

minnus commented:
Thanks for the prompt review!  Yeah, your suggestion worked out well and seems cleaner.

The rational is to allow easier parsing of the euca commands.  We have a variety of tools for accounting and monitoring and it was pretty tedious to parse the formatted output.  Since the euca commands already have the data stored internally in a nice format, it was straightforward to output as json.  We only needed 2 or 3 of the commands but I figured I might as well do them all.

Sure will do docs once you confirm the changes are ok and implementation is stable.

minnus committed:
Fix whitespace

gholms commented on euca2ools/commands/ec2/__init__.py:
```diff
@@ -72,9 +73,19 @@ class EC2Request(AWSQueryRequest, TabifyingMixin):
 
+    ARGS = [Arg('-j', '--json',
+                action='store_const', const='true', route_to=None,
+                help='''Output in json format''')]
```
Let's make this more extensible by making it an option that takes an arg, like `--pretty=json`, so we don't have to clutter the namespace as we add new output formats in the future.

aebruno replied:
This has been fixed. Added pretty=json that accepts a list of choices (only json is supported for now).

gholms commented on euca2ools/commands/ec2/__init__.py:
```diff
@@ -72,9 +73,19 @@ class EC2Request(AWSQueryRequest, TabifyingMixin):
+        if self.args.get('json'):
+            print json.dumps(result, sort_keys=True, indent=2)
+        else:
+            self.print_result_native(result)
```
This needs to handle cases where `print_result_native` is not defined.

aebruno replied:
Renamed this function to print_result_plain and added this to the base class. This should avoid the case where it's not defined? Also added print_result_json so subclasses could potentially override this behavior

gholms commented on euca2ools/commands/ec2/__init__.py:
```diff
@@ -72,9 +73,19 @@ class EC2Request(AWSQueryRequest, TabifyingMixin):
 
+    def print_result(self, result):
+        if self.args.get('json'):
+            print json.dumps(result, sort_keys=True, indent=2)
```
How does this handle types that json.dumps does not have serializers for, such as PaginatedResponses?

gholms commented:
As expected, that makes this _much_ cleaner.  :-)  See the comments above.

I will be out of town for the next week, so I apologize in advance if it takes a while before you get another response.

gholms commented:
How go things, @minnus?

aebruno committed:
Change option name to support multiple output formats

aebruno committed:
Add methods for pretty print output formats

aebruno committed:
Update man pages with pretty option

aebruno committed:
Remove json format check

aebruno commented:
@gholms Sorry for the delay on this PR. I'm continuing @minnus work and would be happy to make any additional fixes necessary to get this accepted. We have several tools for accounting and monitoring our euca cloud that makes use of the pretty=json format. Would be great to see this incorporated into the upstream euca2ools. I attempted to tidy things up according to your previous comments above. Let us know how things look. Thanks!

plessbd commented:
@gholms any update on what we would need for this.  I would also find this useful

plessbd commented:
@gholms any updates on this?  I have been useing the patch to update my local copy and it seems to work, would be awesome if this was standard.
